python/tvm/topi/nn/dense.py

```diff
@@ -51,37 +65,120 @@ def dense(data, weight, bias=None, out_dtype=None, auto_scheduler_rewritten_layo
         assert len(bias.shape) == 1
     if out_dtype is None:
         out_dtype = data.dtype
-    batch, in_dim = data.shape
+    if data_transposed:
+        in_dim, batch = data.shape
+    else:
+        batch, in_dim = data.shape
 
     if auto_scheduler_rewritten_layout:
         # Infer shape for the rewritten layout
         out_dim, red_dim = auto_scheduler.get_shape_from_rewritten_layout(
-            auto_scheduler_rewritten_layout, ["j", "k"]
+            auto_scheduler_rewritten_layout, ["j", "k"] if weight_transposed else ["k", "j"]
         )
         auto_scheduler.remove_index_check(weight)
-    else:
+    elif weight_transposed:
         out_dim, red_dim = weight.shape
+    else:
+        red_dim, out_dim = weight.shape
     assert in_dim == red_dim
 
     k = te.reduce_axis((0, in_dim), name="k")
-    matmul = te.compute(
+    if data_transposed:
```

Comment:
nit: This might be clearer:

```
if (data_transposed, weight_transposed) == (False, False):
    # ...
elif (data_transposed, weight_transposed) == (False, True):
    # ... 
elif (data_transposed, weight_transposed) == (True, False):
    # ...
elif (data_transposed, weight_transposed) == (True, True):
    # ...
```

Also I think it's fine to just use `T_matmul_NT` instead of `T_dense`.